### ISSUE TITLE
ASPLOS 25 | DeepContext: A Context-aware, Cross-platform, and Cross-framework Tool for Performance Profiling and Analysis of Deep Learning Workloads

### ISSUE THREAD

**Issue by woshipapa** [AI-Infra-Team/awesome-papers#187]
DeepContext: A Context-aware, Cross-platform, andCross-framework Tool for Performance Profiling and Analysis of Deep Learning Workloads

Qidong ZhaoNorth Carolina State UniversityRaleigh, NCUnited State[[qzhao24@ncsu.edu](mailto:qzhao24@ncsu.edu)](mailto:qzhao24@ncsu.edu)Hao WuGeorge Mason UniversityFairfax, VAUnited State[[hwu27@gmu.edu](mailto:hwu27@gmu.edu)](mailto:hwu27@gmu.edu)Yueming HaoNorth Carolina State UniversityRaleigh, NCUnited State[[yhao24@ncsu.edu](mailto:yhao24@ncsu.edu)](mailto:yhao24@ncsu.edu)Zilingfeng YeGeorge Mason UniversityFairfax, VAUnited State[[yipzlf@gmail.com](mailto:yipzlf@gmail.com)](mailto:yipzlf@gmail.com)Jiajia LiNorth Carolina State UniversityRaleigh, NCUnited State[[jiajia.li@ncsu.edu](mailto:jiajia.li@ncsu.edu)](mailto:jiajia.li@ncsu.edu)Xu LiuNorth Carolina State UniversityRaleigh, NCUnited State[[xliu88@ncsu.edu](mailto:xliu88@ncsu.edu)](mailto:xliu88@ncsu.edu)Keren ZhouGeorge Mason UniversityFairfax, VAUnited State[[kzhou6@gmu.edu](mailto:kzhou6@gmu.edu)](mailto:kzhou6@gmu.edu)

[Paper](https://arxiv.org/html/2411.02797v1)

## Motivation

<img width="1001" height="282" alt="Image" src="https://github.com/user-attachments/assets/0ebac306-fad7-4ef0-9a71-9f0cd64cc120" />

<img width="467" height="635" alt="Image" src="https://github.com/user-attachments/assets/8b3a9509-1e82-49c4-8b1b-17792f35b731" />

As deep learning workloads become increasingly complex, performance bottlenecks can no longer be understood by looking at isolated CPU or GPU events alone. Modern training and inference pipelines span high-level Python code, deep learning framework operators, C/C++ runtime libraries, and GPU device execution, while also involving sophisticated interactions between CPUs and GPUs. However, existing profiling tools typically expose only part of this stack, making it difficult to attribute low-level performance issues back to the high-level model code and framework semantics that actually triggered them.

At the same time, existing profilers are heavily trace-oriented. They often record every operation in long-running workloads, which leads to massive profile data, high memory and storage overhead, and substantial manual effort during analysis. Although some tools can aggregate metrics after profiling, they usually do so only at the kernel level and fail to preserve enough semantic context for automated diagnosis. As a result, users still need to manually connect hotspots to model structure, operator behavior, and source-level code.

A further challenge is fragmentation across frameworks and hardware platforms. Vendor tools are often tied to specific GPU ecosystems, while framework-specific profilers are limited to individual frameworks such as PyTorch or JAX. This fragmentation raises the learning cost for users and makes it hard to perform consistent performance analysis across different software and hardware stacks. Therefore, the paper is motivated by the need for a profiler that is simultaneously context-aware, compact, and portable across frameworks and platforms.

### The paper’s core question is:

**How can we build a unified profiling system that reconstructs full-stack execution context for deep learning workloads, reduces the burden of massive trace data through structured aggregation and analysis, and remains usable across different frameworks and heterogeneous hardware platforms?**

### This motivation is built on three observations:

**Existing tools lack full-stack semantic context.**  
Most profilers can only show part of the execution process, such as Python, framework events, or low-level kernels, but they do not integrate these layers into a single coherent call path. This makes it hard to identify which high-level operator or source-level code is responsible for a given low-level bottleneck.

**Existing profiling workflows generate large traces but limited optimization insight.**  
Long-running workloads can produce millions of traced events. Existing tools mostly leave interpretation to the user, provide limited online aggregation, and offer little support for automatic issue identification, which makes performance debugging expensive and labor-intensive.

**Existing tools are fragmented across frameworks and platforms.**  
Vendor profilers and framework profilers each cover only part of the ecosystem, which prevents a unified profiling workflow across PyTorch, JAX, Nvidia GPUs, AMD GPUs, and different CPU architectures. The paper argues that an effective profiler should support cross-framework and cross-platform analysis rather than being restricted to one stack.

<img width="438" height="387" alt="Image" src="https://github.com/user-attachments/assets/c219aa99-49ae-4383-bf2e-b9f60ae69322" />

## Solution

To address these issues, the paper proposes **DeepContext**, a profiling framework organized around several coordinated components: **DLMonitor, Profiler, Performance Analyzer, GUI, and a modular adaptation layer for framework/hardware evolution**. Rather than acting as a simple event tracer, DeepContext is designed as a full pipeline that reconstructs cross-layer execution context, attributes CPU/GPU metrics to that context, analyzes the resulting performance structure, and presents optimization hints to users.

### DLMonitor: a shim layer for cross-layer context reconstruction

<img width="464" height="383" alt="Image" src="https://github.com/user-attachments/assets/640d1bbd-809a-440d-a15e-f54f11878e96" />
<img width="461" height="347" alt="Image" src="https://github.com/user-attachments/assets/03a8d89f-9c8c-4fe9-8338-8c8838bde023" />
DeepContext first introduces **DLMonitor**, a shim layer between profilers and deep learning frameworks. DLMonitor intercepts framework operations and GPU APIs, and constructs a unified call path spanning Python code, framework operators, native C/C++ frames, GPU runtime calls, and GPU kernels. For PyTorch, it leverages framework callback interfaces; for JAX, it uses lightweight binary instrumentation to capture operator-level information despite graph compilation and fusion. It also includes mechanisms such as forward/backward operator association, shadow stacks, and call-path caching to preserve semantic context while keeping overhead manageable. This component is the foundation that turns fragmented low-level traces into semantically meaningful execution paths.

### Profiler: metric collection, attribution, and online aggregation

<img width="451" height="350" alt="Image" src="https://github.com/user-attachments/assets/8d9701d6-fa1f-4a22-86bb-e78a3203eed1" />

On top of DLMonitor, DeepContext’s **Profiler** collects both GPU and CPU metrics and attributes them to the reconstructed call paths. GPU metrics are gathered through APIs such as CUPTI and RocTracer, while CPU metrics can come from system calls, perf events, or PAPI. The profiler associates metrics with call paths using correlation IDs and organizes them into a **Calling Context Tree**, where nodes from Python, framework operators, native C/C++, GPU APIs, and kernels are merged according to their own identity rules. Metrics are then aggregated online by sum, min, average, and standard deviation and propagated up the tree. This design reduces profile size while preserving semantic attribution, making long-running workloads more analyzable than flat event traces.

### Performance Analyzer: automated diagnosis on top of the call tree

Once the calling context tree has been built, DeepContext applies an automated **Performance Analyzer**. This analyzer maps low-level instructions back to source code using DWARF information and exposes a flexible Python interface for analysis across program structure, model-level behavior, and operator-level behavior. The paper presents several example analyses, including hotspot identification, excessive small-kernel detection, forward/backward imbalance analysis, fine-grained stall analysis, and CPU latency analysis. The key idea is that DeepContext does not stop at collecting traces; it uses the semantic calling context tree to automatically flag potential optimization opportunities and produce actionable warnings.

### GUI: compact, hierarchical visualization inside the developer workflow

DeepContext also provides a lightweight **GUI** built on the VSCode API. Instead of showing only flat timelines, it visualizes the calling context tree through interactive flame graphs with top-down and bottom-up views, highlights analyzer-detected hotspots, and lets users click through frames to inspect metrics and navigate to source code. This makes the profiling results easier to interpret because the visualization is aligned with hierarchical program structure rather than just chronological event order.

### Modular adaptability: separating framework evolution from hardware evolution

Finally, DeepContext is designed with **modular adaptability** in mind. DLMonitor mainly changes when framework APIs evolve, while the Profiler mainly changes when hardware profiling interfaces evolve. The Performance Analyzer and GUI operate on the abstracted performance data and are therefore largely framework- and hardware-agnostic. This separation makes the system more maintainable as frameworks such as PyTorch and JAX change internally and as profiling support evolves across Nvidia, AMD, x86, and ARM platforms.

<img width="982" height="521" alt="Image" src="https://github.com/user-attachments/assets/1eb55afe-9f17-4ec2-a3ea-9556a2f20d0b" />

<img width="495" height="999" alt="Image" src="https://github.com/user-attachments/assets/8c2fb3cf-f69c-47a3-bfc9-d793c268a2f4" />

**Comment by qzweng** [comment]
1st author: https://www.linkedin.com/in/qidong-zhao/ Google Software Engineer
3rd author: https://www.linkedin.com/in/haoyueming/
Meta: https://github.com/meta-pytorch/tritonparse
